crates/ruff/src/rules/eradicate/rules.rs

```diff
@@ -9,20 +9,20 @@ use crate::settings::{flags, Settings};
 use crate::source_code::Locator;
 use crate::violation::AlwaysAutofixableViolation;
 
-define_violation!(
-    /// ## What it does
-    /// Checks for commented-out Python code.
-    ///
-    /// ## Why is this bad?
-    /// Commented-out code is dead code, and is often included inadvertently.
-    /// It should be removed.
-    ///
-    /// ## Example
-    /// ```python
-    /// # print('foo')
-    /// ```
-    pub struct CommentedOutCode;
-);
+/// ## What it does
+/// Checks for commented-out Python code.
+///
+/// ## Why is this bad?
+/// Commented-out code is dead code, and is often included inadvertently.
+/// It should be removed.
+///
+/// ## Example
+/// ```python
+/// # print('foo')
+/// ```
+#[violation]
```

Comment:
Nice! So we're using an attribute macro rather than a function-like macro. This does strike me as more intuitive. 

What would be the motivation to move from an attribute macro to a derive macro? How would that differ?

Reply:
A derive macro can only define the contents of an impl block. So we would first define a trait for, either changing `Violation` or a new `ViolationExplanation` and then use it while also adding the other derives manually:

```rust
#[derive(Violation, Debug, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
/// # Why is this bad
/// Has a rather stale taste
struct EatLaundry;
```

We could also go all the way refactoring and do a proper trait hierarchy. Note that we still need two proc macros and traits because we need to parse one method and create another.

```rust
trait ViolationMessage {
    fn message(&self) -> String;
    fn derived_messages(&self) -> Vec<&'static str>;
    fn autofix_title(&self) -> String {
        ""
    }
}

trait Violation: ViolationMessage + Debug + PartialEq + Eq + serde::Serialize + serde::Deserialize {
    fn explanation() -> String;
}

#[derive(Violation, Debug, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
/// # Why is this bad
/// Has a rather stale taste
struct EatLaundry;

#[violation_messages] // Adds derived_messages()
impl ViolationMessage for EatLaundry {
    fn message(&self) -> String {
        format!("Found laundry-eating behaviour")
    }
    fn autofix_title(&self) -> String {
        "Makes the code munch on tokens instead".to_string()
    }
}
```

A more elegant alternative would be what [thiserror does](https://docs.rs/thiserror/latest/thiserror/), but that would be an awful lot of work for just an internal helper macro (unless [miette](https://docs.rs/miette/latest/miette/) or something gets us that for free, but i haven't checked and i'd be surprised if there was a perfect fit already existing):

```rust
use thiserror::Error;

#[derive(Error, Debug)]
pub enum DataStoreError {
    #[error("data store disconnected")]
    Disconnect(#[from] io::Error),
    #[error("the data for key `{0}` is not available")]
    Redaction(String),
    #[error("invalid header (expected {expected:?}, found {found:?})")]
    InvalidHeader {
        expected: String,
        found: String,
    },
    #[error("unknown data store error")]
    Unknown,
}
```